src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py

```diff
@@ -15690,8 +15690,120 @@ def dt_round(
+                slice_length *= 24
+                slice_unit = "hour"
+            else:
+                f"Snowpark pandas 'Series.dt.round' method doesn't support setting 'freq' parameter with '{slice_unit}' unit"
+            return slice_length, slice_unit
+
+        if slice_length % 2 == 1:
+            slice_length, slice_unit = down_level_freq(slice_length, slice_unit)
+        half_slice_length = int(slice_length / 2)
+        base_column = col(internal_frame.data_column_snowflake_quoted_identifiers[0])
+
+        # Second, we determine whether floor represents an even number of slices.
+        # To do so, we must divide the number of epoch seconds in it over the number
+        # of epoch seconds in one slice. This way, we can get the number of slices.
+
+        floor_column = builtin("time_slice")(
+            base_column, slice_length, slice_unit, "START"
+        )
+        ceil_column = builtin("time_slice")(
+            base_column, slice_length, slice_unit, "END"
+        )
+
+        def slice_length_when_unit_is_second(slice_length: int, slice_unit: str) -> int:
+            while slice_unit != "second":
+                slice_length, slice_unit = down_level_freq(slice_length, slice_unit)
+            return slice_length
+
+        floor_epoch_seconds_column = builtin("extract")("epoch_second", floor_column)
+        floor_num_slices_column = cast(
+            floor_epoch_seconds_column
+            / pandas_lit(slice_length_when_unit_is_second(slice_length, slice_unit)),
+            IntegerType(),
+        )
+
+        # Now that we know the number of slices, we can check if they are even or odd.
+
+        floor_is_even = (floor_num_slices_column % pandas_lit(2)).equal_null(
+            pandas_lit(0)
+        )
+
+        # Accordingly, we can decide if the round column should be the floor or ceil
+        # of the slice.
+
+        round_column_if_half_point = iff(floor_is_even, floor_column, ceil_column)
+
+        # In case the date value is not at half point of the slice, then we shift it
+        # by half a slice, and take the floor from there.
+
+        base_plus_half_slice_column = dateadd(
+            slice_unit, pandas_lit(half_slice_length), base_column
+        )
+        round_column_if_not_half_point = builtin("time_slice")(
+            base_plus_half_slice_column, slice_length, slice_unit, "START"
+        )
+
+        # The final expression for the round column.
+
+        round_column = iff(
+            base_plus_half_slice_column.equal_null(ceil_column),
+            round_column_if_half_point,
+            round_column_if_not_half_point,
+        )
+
+        internal_frame = internal_frame.append_column(
+            internal_frame.data_column_pandas_labels[0], round_column
+        )
+
+        return SnowflakeQueryCompiler(
+            InternalFrame.create(
+                ordered_dataframe=internal_frame.ordered_dataframe,
+                data_column_pandas_labels=[None],
```

Comment:
We have to carry the `data_column_pandas_labels` over to the result frame. Currently this works in pandas but not in Snowpark pandas:

```python
 assert pd.Series(pd.date_range('1/1/2018 11:59:00', periods=3, freq='min'), name='z').dt.round("h").name == 'z'
```

Can we use `InternalFrame.update_snowflake_quoted_identifiers_with_expressions` instead?

Reply:
Good catch. Fixed.